File: rsocket-java-ffi/examples/PerformanceBenchmark.java
import com.rsocket.rust.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicLong;

public class PerformanceBenchmark {
    public static void main(String[] args) {
        System.out.println("RSocket Java FFI Performance Benchmark");
        System.out.println("======================================");
        
        RSocket.init();
        
        final int numRequests = 1000;
        final int messageSize = 1024;
        
        byte[] message = new byte[messageSize];
        for (int i = 0; i < messageSize; i++) {
            message[i] = (byte) (i % 256);
        }
        
        System.out.printf("Sending %d requests with %d byte payloads...%n", numRequests, messageSize);
        
        try (RSocketClient client = new RSocketClient()) {
            client.connectTcp("127.0.0.1:7878");
            
            long startTime = System.currentTimeMillis();
            CountDownLatch latch = new CountDownLatch(numRequests);
            AtomicLong successCount = new AtomicLong(0);
            AtomicLong errorCount = new AtomicLong(0);
            
            for (int i = 0; i < numRequests; i++) {
                try (Payload payload = Payload.fromBytes(message, null)) {
                    CompletableFuture<Payload> future = client.requestResponse(payload);
                    
                    future.whenComplete((response, throwable) -> {
                        if (throwable != null) {
                            errorCount.incrementAndGet();
                        } else {
                            successCount.incrementAndGet();
                            if (response != null) {
                                response.close();
                            }
                        }
                        latch.countDown();
                    });
                }
            }
            
            latch.await();
            long duration = System.currentTimeMillis() - startTime;
            
            System.out.println("\nBenchmark Results:");
            System.out.println("Duration: " + duration + "ms");
            System.out.println("Successful requests: " + successCount.get());
            System.out.println("Failed requests: " + errorCount.get());
            System.out.printf("Requests per second: %.2f%n", (double) successCount.get() / (duration / 1000.0));
            System.out.printf("Average latency: %.2fms%n", (double) duration / successCount.get());
            
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: rsocket-go-ffi/examples/performance-benchmark.go
package main

import (
    "fmt"
    "log"
    "sync"
    "time"
    
    rsocket "../go"
)

func main() {
    fmt.Println("RSocket Go FFI Performance Benchmark")
    fmt.Println("====================================")
    
    if err := rsocket.Init(); err != nil {
        log.Fatalf("Failed to initialize RSocket: %v", err)
    }
    
    client := rsocket.NewClient()
    if client == nil {
        log.Fatal("Failed to create client")
    }
    
    if err := client.ConnectTCP("127.0.0.1:7878"); err != nil {
        log.Fatalf("Failed to connect: %v", err)
    }
    
    const numRequests = 1000
    const messageSize = 1024
    
    message := make([]byte, messageSize)
    for i := range message {
        message[i] = byte(i % 256)
    }
    
    fmt.Printf("Sending %d requests with %d byte payloads...\n", numRequests, messageSize)
    
    start := time.Now()
    var wg sync.WaitGroup
    var successCount int64
    var errorCount int64
    var mu sync.Mutex
    
    for i := 0; i < numRequests; i++ {
        wg.Add(1)
        
        payload := rsocket.NewPayloadFromBytes(message, nil)
        if payload == nil {
            log.Fatal("Failed to create payload")
        }
        
        err := client.RequestResponse(payload, func(response *rsocket.Payload, err error) {
            defer wg.Done()
            
            mu.Lock()
            if err != nil {
                errorCount++
            } else {
                successCount++
            }
            mu.Unlock()
        })
        
        if err != nil {
            wg.Done()
            mu.Lock()
            errorCount++
            mu.Unlock()
        }
    }
    
    wg.Wait()
    duration := time.Since(start)
    
    fmt.Printf("\nBenchmark Results:\n")
    fmt.Printf("Duration: %v\n", duration)
    fmt.Printf("Successful requests: %d\n", successCount)
    fmt.Printf("Failed requests: %d\n", errorCount)
    fmt.Printf("Requests per second: %.2f\n", float64(successCount)/duration.Seconds())
    fmt.Printf("Average latency: %v\n", duration/time.Duration(successCount))
}
